### DISCUSSION TITLE
Canvas vs PNG composition, bug or misusage?

### DISCUSSION THREAD

**Discussion by mxi** [ImageMagick/ImageMagick#3967] [Help] [answered]
I don't want to submit a bug report just in case this is the correct behavior.

I'm looking to make a quad circle and the best way I figured to go about this was to use blending modes. With PNG images, the `-compose In -composite` seemed to work just fine, but with transparent canvases, regardless of stacking, it does not appear to oblige the blending mode. 

I have created a test script to demonstrate the behavior:
```sh
#!/bin/sh

magick --version

echo ""

# works as expected
magick -size 500x500 canvas:transparent -fill black -draw "circle 250,250,500,250" circle.png
magick -size 500x500 canvas:transparent -fill black -draw "rectangle 250,250,500,500" rect.png
magick circle.png -colorspace RGB -identify rect.png -colorspace RGB -identify -compose In -composite comp.png


# does not work as expected, bug or misunderstanding?
magick -size 500x500 -depth 8 -fill black \
	\( \
		canvas:transparent \
		-colorspace RGB \
		-identify \
		-draw "circle 250,250,500,250" \
		canvas:transparent \
		-colorspace RGB \
		-identify \
		-draw "rectangle 250,250,500,500" \
		-compose In \
		-composite \
	\) \
	stack-comp_with-compose.png

## clipping mask did not work
# magick -size 500x500 -depth 8 -fill black \
# 	\( \
# 		canvas:transparent \
# 		-colorspace RGB \
# 		-identify \
# 		-draw "rectangle 250,250,500,500" \
# 		-clip-mask \
# 		canvas:transparent \
# 		-colorspace RGB \
# 		-identify \
# 		-draw "circle 250,250,500,250" \
# 	\) \
# 	stack-comp.png

# region seems to work
magick -size 500x500 -depth 8 -fill black \
	\( \
		canvas:transparent \
		-region '250x250+250+250' \
		-colorspace RGB \
		-identify \
		-draw "circle 250,250,500,250" \
	\) \
	stack-comp_with-region.png
```
with the following output:
```
Version: ImageMagick 7.1.0-5 Q16 x86_64 2021-07-18 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): fontconfig freetype jbig jng jp2 jpeg lzma openexr pangocairo png tiff webp x xml zlib

circle.png PNG 500x500 500x500+0+0 8-bit Gray 7549B 0.000u 0:00.002
circle.png[0] PNG 500x500 500x500+0+0 8-bit Gray 7549B 0.010u 0:00.002
rect.png[0] PNG 500x500 500x500+0+0 8-bit Gray 1245B 0.010u 0:00.002
canvas:transparent=>transparent CANVAS 500x500 500x500+0+0 8-bit RGB 0.020u 0:00.004
canvas:transparent=>transparent[0] CANVAS 500x500 500x500+0+0 8-bit RGB 0.060u 0:00.013
canvas:transparent=>transparent[0] CANVAS 500x500 500x500+0+0 8-bit RGB 0.020u 0:00.004
canvas:transparent=>transparent CANVAS 500x500 500x500+0+0 8-bit RGB 0.020u 0:00.008
```
The images produced are:

- PNG composition:
  ![comp](https://user-images.githubusercontent.com/26512240/126425755-fabfee87-f210-486b-aae1-d7a8cde3c2d8.png)

- Canvas composition:
  ![stack-comp_with-compose](https://user-images.githubusercontent.com/26512240/126425821-e6ad3f63-b5ed-4c9b-9520-4322c6c10a35.png)

- Region:
  ![stack-comp_with-region](https://user-images.githubusercontent.com/26512240/126425878-23d5e541-1a48-452b-9678-e9a8c8469258.png)

Appreciate any help, cheers!

**Comment by fmw42** [comment]
I am unsure what you want. What is a quad circle?  Can you post an example of the output you want or a drawing of what you want?

**Reply by mxi** [reply]
My apologies, I just woke up and the answer by snlbgo was the one I was looking for.

I'm blanking on the term but I just want a quarter circle instead of a full circle but without using the `path` draw command. In any case, this is resolved now. Appreciate your time.

**Comment by snibgo** [comment, marked as answer]
> does not work as expected, bug or misunderstanding?

magick -size 500x500 -depth 8 -fill black \
	\( \
		canvas:transparent \
		-colorspace RGB \
		-identify \
		-draw "circle 250,250,500,250" \
		canvas:transparent \
		-colorspace RGB \
		-identify \
		-draw "rectangle 250,250,500,500" \
		-compose In \
		-composite \
	\) \
	stack-comp_with-compose.png

Misunderstanding. Most IM operationsm such as "-draw", operate on _all_ the images in the current list. So you draw a rectangle on both the images, not just the second image.

To restrict that draw to just the second image, use parentheses:
```
magick -size 500x500 -depth 8 -fill black \
	canvas:transparent \
	-colorspace RGB \
	-identify \
	-draw "circle 250,250,500,250" \
	\( \
		canvas:transparent \
		-colorspace RGB \
		-identify \
		-draw "rectangle 250,250,500,500" \
	\) \
	-compose In \
	-composite \
	s2.png
```
I have removed your outermost parentheses because they are pointless.

**Reply by mxi** [reply]
Ahh ok. This makes a lot more sense. Appreciate it.
